Cython/Compiler/Nodes.py

```diff
@@ -1267,16 +1266,27 @@ def _analyse_template_types(self, env, base_type):
                     error(template_node.pos, "unknown type in template argument")
                     ttype = error_type
                 # For Python generics we can be a bit more flexible and allow None.
-            elif require_python_types and not ttype.is_pyobject or require_optional_types and not ttype.can_be_optional():
+            template_types.append(ttype)
+
+        if base_type.python_type_constructor_name:
+            base_type.contains_none = PyrexTypes.py_none_type in template_types
+            require_optional_types = base_type.allows_none()
```

Comment:
I think we should only set `contains_none` for `Union[]` declarations. This seems broader.